File: python-algo/leetcode/problem_127.py
"""127. Word Ladder
https://leetcode.com/problems/word-ladder/

Given two words (beginWord and endWord), and a dictionary's word list,
find the length of shortest transformation sequence from beginWord to endWord, such that:

    Only one letter can be changed at a time.
    Each transformed word must exist in the word list.

Note:

    Return 0 if there is no such transformation sequence.
    All words have the same length.
    All words contain only lowercase alphabetic characters.
    You may assume no duplicates in the word list.
    You may assume beginWord and endWord are non-empty and are not the same.

Example 1:

Input:
beginWord = "hit",
endWord = "cog",
wordList = ["hot","dot","dog","lot","log","cog"]

Output: 5

Explanation: As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog",
return its length 5.

Example 2:

Input:
beginWord = "hit"
endWord = "cog"
wordList = ["hot","dot","dog","lot","log"]

Output: 0

Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
"""
from typing import List


def ladder_length(begin_word: str, end_word: str, word_list: List[str]) -> int:
    # BFS
    word_set = set(word_list)
    if end_word not in word_set:
        return 0
    word_len = len(end_word)
    queue = [begin_word]
    ans = 1
    while queue:
        n = len(queue)
        for i in range(n):
            cur_word = queue.pop(0)
            if cur_word == end_word:
                return ans
            for j in range(word_len):
                for k in range(26):
                    char = chr(ord('a') + k)
                    if char != cur_word[j]:
                        new_word = cur_word[:j] + char + cur_word[j + 1:]
                        if new_word in word_set:
                            word_set.remove(new_word)
                            queue.append(new_word)
        ans += 1
    return 0


def ladder_length_2(begin_word: str, end_word: str, word_list: List[str]) -> int:
    # Bi-BFS
    def bfs(q1: List[str], q2: List[str]) -> int:
        nonlocal ans
        while q1 and q2:
            n1, n2 = len(q1), len(q2)
            if n1 > n2:
                return bfs(q2, q1)
            for i in range(n1):
                cur_word = q1.pop(0)
                for j in range(word_len):
                    for k in range(26):
                        char = chr(ord('a') + k)
                        if char != cur_word[j]:
                            new_word = cur_word[:j] + char + cur_word[j + 1:]
                            if new_word in q2:
                                return ans + 1
                            if new_word in word_set:
                                word_set.remove(new_word)
                                q1.append(new_word)
            ans += 1
        return 0

    word_set = set(word_list)
    if end_word not in word_set:
        return 0
    word_len = len(end_word)
    ans = 1
    return bfs([begin_word], [end_word])

File: java-algo/src/main/java/com/leetcode/Problem127.java
package com.leetcode;

import java.util.*;

/**
 * 127. Word Ladder
 * https://leetcode.com/problems/word-ladder/
 * <p>
 * Given two words (beginWord and endWord), and a dictionary's word list,
 * find the length of shortest transformation sequence from beginWord to endWord, such that:
 * <p>
 * Only one letter can be changed at a time.
 * Each transformed word must exist in the word list.
 * <p>
 * Note:
 * <p>
 * Return 0 if there is no such transformation sequence.
 * All words have the same length.
 * All words contain only lowercase alphabetic characters.
 * You may assume no duplicates in the word list.
 * You may assume beginWord and endWord are non-empty and are not the same.
 * <p>
 * Example 1:
 * <p>
 * Input:
 * beginWord = "hit",
 * endWord = "cog",
 * wordList = ["hot","dot","dog","lot","log","cog"]
 * <p>
 * Output: 5
 * <p>
 * Explanation: As one shortest transformation is "hit" -> "hot" -> "dot" -> "dog" -> "cog",
 * return its length 5.
 * <p>
 * Example 2:
 * <p>
 * Input:
 * beginWord = "hit"
 * endWord = "cog"
 * wordList = ["hot","dot","dog","lot","log"]
 * <p>
 * Output: 0
 * <p>
 * Explanation: The endWord "cog" is not in wordList, therefore no possible transformation.
 */
public class Problem127 {

    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (wordSet.size() == 0 || !wordSet.contains(endWord))
            return 0;
        // BFS
        Queue<String> queue = new LinkedList<>();
        queue.add(beginWord);
        int ans = 1;
        while (!queue.isEmpty()) {
            int n = queue.size();
            for (int i = 0; i < n; i++) {
                String word = queue.poll();
                if (endWord.equals(word))
                    return ans;
                wordSet.remove(word);
                for (int j = 0; j < word.length(); j++) {
                    for (char c = 'a'; c <= 'z'; c++) {
                        String newWord = word.substring(0, j) + c + word.substring(j + 1);
                        if (wordSet.contains(newWord)) {
                            queue.add(newWord);
                        }
                    }
                }
            }
            ans++;
        }
        return 0;
    }

    public int ladderLength2(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (wordSet.size() == 0 || !wordSet.contains(endWord))
            return 0;
        int ans = 1;
        // Bi-BFS
        // be aware that use HashSet rather than Queue here.
        Set<String> beginSet = new HashSet<>();
        Set<String> endSet = new HashSet<>();
        beginSet.add(beginWord);
        endSet.add(endWord);
        while (!beginSet.isEmpty() && !endSet.isEmpty()) {
            // swap if needed to keep beginSet to be the little one.
            if (beginSet.size() > endSet.size()) {
                Set<String> temp = beginSet;
                beginSet = endSet;
                endSet = temp;
            }
            Set<String> nextBeginSet = new HashSet<>();
            for (String word : beginSet) {
                if (endSet.contains(word))
                    return ans;
                wordSet.remove(word);
                for (int j = 0; j < word.length(); j++) {
                    for (char c = 'a'; c <= 'z'; c++) {
                        String newWord = word.substring(0, j) + c + word.substring(j + 1);
                        if (wordSet.contains(newWord)) {
                            nextBeginSet.add(newWord);
                        }
                    }
                }
            }
            beginSet = nextBeginSet;
            ans++;
        }
        return 0;
    }
}
